server/src/main/java/com/linecorp/centraldogma/server/internal/api/CredentialServiceV1.java

```diff
@@ -0,0 +1,115 @@
+import com.linecorp.centraldogma.internal.api.v1.PushResultDto;
+import com.linecorp.centraldogma.server.command.CommandExecutor;
+import com.linecorp.centraldogma.server.internal.api.auth.RequiresReadPermission;
+import com.linecorp.centraldogma.server.internal.api.auth.RequiresWritePermission;
+import com.linecorp.centraldogma.server.internal.storage.project.ProjectApiManager;
+import com.linecorp.centraldogma.server.mirror.MirrorCredential;
+import com.linecorp.centraldogma.server.storage.project.Project;
+import com.linecorp.centraldogma.server.storage.repository.MetaRepository;
+
+/**
+ * Annotated service object for managing credential service.
+ */
+@ProducesJson
+@RequiresWritePermission(repository = Project.REPO_META)
+@ExceptionHandler(HttpApiExceptionHandler.class)
+public class CredentialServiceV1 extends AbstractService {
+
+    private final ProjectApiManager projectApiManager;
+
+    public CredentialServiceV1(ProjectApiManager projectApiManager, CommandExecutor executor) {
+        super(executor);
+        this.projectApiManager = projectApiManager;
+    }
+
+    /**
+     * GET /projects/{projectName}/credentials
+     *
+     * <p>Returns the list of the credentials in the project.
+     */
+    @RequiresReadPermission(repository = Project.REPO_META)
+    @Get("/projects/{projectName}/credentials")
+    public CompletableFuture<List<MirrorCredential>> listCredentials(@Param String projectName) {
+        return metaRepo(projectName).credentials();
+    }
+
+    /**
+     * GET /projects/{projectName}/credentials/{id}
+     *
+     * <p>Returns the credential for the ID in the project.
+     */
+    @RequiresReadPermission(repository = Project.REPO_META)
+    @Get("/projects/{projectName}/credentials/{id}")
+    public CompletableFuture<MirrorCredential> getCredentialById(@Param String projectName, @Param String id) {
+        return metaRepo(projectName).credential(id);
+    }
+
+    /**
+     * POST /projects/{projectName}/credentials
+     *
+     * <p>Creates a new credential.
+     */
+    @Post("/projects/{projectName}/credentials")
+    @ConsumesJson
+    @StatusCode(201)
+    public CompletableFuture<PushResultDto> createCredential(@Param String projectName,
+                                                             MirrorCredential credential, Author author) {
+        return createOrUpdate(projectName, credential, author, false);
+    }
+
+    /**
+     * PUT /projects/{projectName}/credentials
+     *
+     * <p>Update the existing credential.
+     */
+    @Put("/projects/{projectName}/credentials")
+    @ConsumesJson
+    public CompletableFuture<PushResultDto> updateCredential(@Param String projectName,
+                                                             MirrorCredential credential, Author author) {
+        return createOrUpdate(projectName, credential, author, true);
+    }
+
+    private CompletableFuture<PushResultDto> createOrUpdate(String projectName,
```

Comment:
Question) Is there a reason we separate create from update? (e.g. is it easier to express in the UI?)  Given the distributed nature of CD, I think I usually prefer that we provide an upsert method instead which is why I'm asking. (e.g. the semantics of `update` may not hold if two concurrent requests go through).

Reply:
I wanted to implement a more restful API.
- `create` needs `POST` method and `201 Created` status is returned.
- `update` uses `PUT` method to replace the whole content and 200 status will be returned instead.

Aside from the REST endpoint, the internal goes through the same code path.

Reply:
> Aside from the REST endpoint, the internal goes through the same code path.

Understood. I personally I prefer simplicity but I think it's fine to keep as is then.

Reply:
`update` is not necessary for this API, but a consistent convention for all REST endpoints would be better. 

Users may get confused if only some resources provide the PUT operation.
As REST API continues to be added to CD, there will be a high possibility of using PUT in the future.
- POST is a non-idempotent operation but `PUT` is so retryable. 
- We may add more validations to `PUT` if some fields should be immutable.

Reply:
> a consistent convention for all REST endpoints would be better.

I'm not strong on the definition/usefulness of REST. I just wanted to make clear the reasoning for adding this logic in case we need to revisit later.